### ISSUE TITLE
Entity with string "status" field breaks React reducers

### ISSUE THREAD

**Issue by ruddell** [jhipster/generator-jhipster#7526]
##### **Overview of the issue**
If an entity contains a `string` field named `status`, webpack compilation of the reducer fails with an error:

```
ERROR in /private/tmp/jh/react/src/main/webapp/app/entities/foo/foo.reducer.ts
(103,12): Type 'AxiosPromise<any>' cannot be converted to type 'Promise<IFoo>'.

ERROR in /private/tmp/jh/react/src/main/webapp/app/entities/foo/foo.reducer.ts
(103,12): Type 'AxiosPromise<any>' cannot be converted to type 'Promise<IFoo>'.
  Types of property 'then' are incompatible.
    Type '<TResult1 = AxiosResponse<any>, TResult2 = never>(onfulfilled?: (value: AxiosResponse<any>) => TR...' is not comparable to type '<TResult1 = IFoo, TResult2 = never>(onfulfilled?: (value: IFoo) => TResult1 | PromiseLike<TResult...'.
      Types of parameters 'onfulfilled' and 'onfulfilled' are incompatible.
        Types of parameters 'value' and 'value' are incompatible.
          Type 'AxiosResponse<any>' is not comparable to type 'IFoo'.
            Types of property 'status' are incompatible.
              Type 'number' is not comparable to type 'string'.

ERROR in /private/tmp/jh/react/src/main/webapp/app/entities/foo/foo.reducer.ts
(110,14): Type 'AxiosPromise<any>' cannot be converted to type 'Promise<IFoo>'.
```

##### **Motivation for or Use Case**
This should work out of the box

##### **Reproduce the error**
Generate a React project and the entity with the following JDL:
```
entity Foo {
    status String
}
```

##### **Suggest a Fix**
Not sure why this happens, it doesn't affect entities without a `string` `status` field.  The part of the error that says `Type 'number' is not comparable to type 'string'` makes me guess it's a conflict between the http status field and the entity status field?

##### **JHipster Version(s)**
v5-beta
master 198234c9e4da659c69eb37edc8987d48e7219b0c

##### **JHipster configuration**

<details>
<summary>.yo-rc.json file</summary>
<pre>
{
  "generator-jhipster": {
    "promptValues": {
      "packageName": "com.mycompany.myapp",
      "nativeLanguage": "en"
    },
    "jhipsterVersion": "5.0.0-beta.0",
    "baseName": "react",
    "reactive": false,
    "packageName": "com.mycompany.myapp",
    "packageFolder": "com/mycompany/myapp",
    "serverPort": "8080",
    "authenticationType": "jwt",
    "cacheProvider": "ehcache",
    "enableHibernateCache": true,
    "websocket": false,
    "databaseType": "sql",
    "devDatabaseType": "h2Disk",
    "prodDatabaseType": "mysql",
    "searchEngine": false,
    "messageBroker": false,
    "serviceDiscoveryType": false,
    "buildTool": "maven",
    "enableSwaggerCodegen": false,
    "jwtSecretKey": "replaced-by-jhipster-info",
    "clientFramework": "react",
    "useSass": false,
    "clientPackageManager": "yarn",
    "applicationType": "monolith",
    "testFrameworks": [],
    "jhiPrefix": "jhi",
    "enableTranslation": true,
    "nativeLanguage": "en",
    "languages": [
      "en"
    ]
  }
}
</pre>
</details>

##### **Entity configuration(s) `entityName.json` files generated in the `.jhipster` directory**
```
entity Foo {
    status String
}
```

##### **Environment and Tools**

<details>
<summary>Environment and Tools</summary>

<pre>

java version "1.8.0_144"
Java(TM) SE Runtime Environment (build 1.8.0_144-b01)
Java HotSpot(TM) 64-Bit Server VM (build 25.144-b01, mixed mode)

git version 2.17.0

node: v8.9.3

npm: 5.6.0

yeoman: 2.0.0

yarn: 1.3.2

Docker version 18.04.0-ce, build 3d479c0

docker-compose version 1.21.0, build 5920eb0

</pre>
</details>

- [X] Checking this box is mandatory (this is just to show you read everything)

**Comment by agaspardcilia** [comment]
It looks like the the http status is stored as `status` in an `AxiosPromise`. It creates a type error between the entity's `status` that is a string and the promise's `status` that is a number.

**Comment by agaspardcilia** [comment]
After looking at Axios source code, I found what was causing this in `http.js`: 
```javascript
var response = {
        status: res.statusCode,
        statusText: res.statusMessage,
        headers: res.headers,
        config: config,
        request: lastRequest
};
```
All this fields can't be used without creating types errors with our current implementation.
@deepu105 What's the best way to deal with this ?

**Comment by deepu105** [comment]
For now, I guess we can add them as reserved keywords like we do for
Angular. But what I dont understand is how is the response holding the
entity, isnt it in body?

May be we can figure out some trick when typing these

Thanks & Regards,
Deepu

**Comment by agaspardcilia** [comment]
Yes, it supposed to be stored in `response.data`. We could also remove `ICrudGetAction<IFoo>` and `Promise<IFoo>` from the reducer action as it doesn't seem to break anything while we try to find a better solution to this problem. wdyt ?

**Comment by deepu105** [comment]
can you make a PR with your suggestion, let me take a look

Thanks & Regards,
Deepu

**Comment by jdubois** [comment]
I'm adding a reserved keywords list for React in JHipster Core (ping @MathieuAA ) and I will link my commit here.

- In the JDL, as the model isn't specific for React or Angular, we will use both reserved keywords lists
- In the Yeoman prompts, depending on your selected client framework, we will use the specific reserved keywords list

This will not close this ticket, as it would be better not to have a reserved keywords list for this, and there should be a cleaner solution. But at least this will prevent users to get this error for the moment, and in the long term it's good to have a reserved keywords list ready for React users, like we have for Angular users.

**Comment by agoncal** [comment]
Argh !!! This is a tough one. We have `status` all over the place in our model: on our invoices, quotations, documents... I wanted to re-generate our JDL with JHipster 5 Beta 1 and got 

```
The JDL is being parsed.
IllegalNameException: The name 'status' is a reserved keyword and can not be used as entity field name.
```

Are you sure to add `status` as a keyword ? JHipster aggregates several technologies and all of them have their keyword. So I think we should be careful about the keywords added in JHipster.

**Comment by jdubois** [comment]
Yes that definitely sucks, I need to check this.

**Comment by deepu105** [comment]
It would be nice to get rid of the reserved keywords, and find some other
way to do this.

Antonio, any ideas? What about prefixing them with something

**Comment by jdubois** [comment]
As we already prefix some stuff (like database tables), we should use a prefix when there is a reserved keyword, instead of forbidding it. Not sure what we should do, maybe something very simple like a "_"?

**Comment by bertwin** [comment]
I believe this line:
https://github.com/jhipster/react-jhipster/blob/b48b95ffabdb004737837de725a5c09af707b156/src/type/redux-action.type.ts#L3

should be changed to either:
```
payload: Promise<{data: T}>;
```
or
```
payload: AxiosPromise<T>;
```

and in all reducers this should be changed:
https://github.com/jhipster/generator-jhipster/blob/ddbdba9d89d5f08b8214f3b998a987119b5db27d/generators/entity-client/templates/react/src/main/webapp/app/entities/entity.reducer.ts.ejs#L208

to 

```
payload: axios.get<I<%= entityReactName %>>(`${requestUrl}${sort ? '&' : '?'}cacheBuster=${new Date().getTime()}`) 
```

**Comment by deepu105** [comment]
@bertwin your suggestion is good, would you like to add it to your PR for redux types #7578

**Comment by deepu105** [comment]
fixed by #7578 thanks to @bertwin

**Comment by agoncal** [comment]
@deepu105 I talked the other day with @jdubois about this. On the JPA side, that's pretty easy, it's just a matter of prefixing the name and using `@Column` (JHipster already does it with `@Table`):

```
@Column(name = "jhi_status")
private String status;
```

But not all the technologies/languages have this easy way to override.

**Comment by jdubois** [comment]
Yes, at least for the database keywords, it's quite easy: we just prefix the columns/tables and then modify the mapping accordingly.

It's quite a bit of work to do, but that would already solve a big part of the issue. Then, I guess people coding in Java will understand that "package" is a reserved keyword, and it's not that bad.

**Comment by bertwin** [comment]
@jdubois @agoncal i haven't tested it, but in theory these reserved words are not needed anymore:
```
 REACT: [
    'STATUS', 'STATUSTEXT', 'HEADERS', 'CONFIG', 'REQUEST'
  ],
```

**Comment by jdubois** [comment]
Oh yes thanks @bertwin - I got confused because we also added some new reserved keywords for MySQL 8, but the real original issue is the one you refer too -> I'm testing this and removing those if possible!

Give me 10 minutes :-)

**Comment by jdubois** [comment]
OK, maybe a bit more :-)
I'm reopening this, I don't know what happens but I guess the master build is currently broken

**Comment by jdubois** [comment]
Ping @MathieuAA it looks like with the current JHipster Core master branch, running `yarn test` on generator-jhipster fails - I don't understand why this wouldn't be failing with Travis?

**Comment by deepu105** [comment]
@jdubois this specific issue is already fixed in master. Did you encounter it again?
Also when I did my last commit today master was green

**Comment by MathieuAA** [comment]
Interesting. Let me have a look.

**Comment by jdubois** [comment]
@deepu105 @MathieuAA I added "STATUS" (and its friends) as a reserved keyword in JHipster Core, so even if it's solved in the generator, you still can't use it. So I would just like to remove it from JHipster Core (I think I'm going to do the change now, it's trivial, and I'll link it here), and then we need a new release of JHipster Core.

**Comment by jdubois** [comment]
Hold on @MathieuAA it might be my fault.... Bad installation with Homebrew as I'm also working to have Node 10 working

**Comment by jdubois** [comment]
Nope sorry @MathieuAA it's not my fault! Here is the error:

```
1) JHipster generator import jdl
       imports a JDL app and entities
         "before all" hook:
     Uncaught Error: ERROR! Error while parsing applications and entities from the JDL Error: The application must be valid in order to be converted.
Errors: No application

**Comment by jdubois** [comment]
I'm closing this as I confirm that the React keywords are fixed, so the original issue is fixed, once jhipster-core is released of course!!!
@MathieuAA don't hesitate to ping me if you need help, also I don't understand how this could have passed Travis

**Comment by bertwin** [comment]
Is there a test that tests generator-jhipster with the latest jhipster-core?
It would have to yarn link it, right?

**Comment by jdubois** [comment]
Or the other way around, there should be a test in JHipster Core that tests with the latest generator from master. Not sure this exists, and it should.

**Comment by MathieuAA** [comment]
The only logical place where there should be a test is in the generator (and there are a few, actually). Anyway, I think I'm done with fixing v2.

**Comment by deepu105** [comment]
Yes there is already a test for import jdl but it uses released version of
jh core and not master

**Comment by MathieuAA** [comment]
Basically, after I'm done with my changes, the subgen will only:
  - call one class from JCore (that will take care of **everything**, but will not make one's coffee _yet_)
    - waaaaaay easier for me to test everything
    - no need for anyone to call a ton of classes from JCore to do one little thing
  - call the other subgens (entity & app) when needed

I don't expect the tests for the JDL import to fail, but I may change them (shouldn't be necessary as everything will be done inside JCore except calling the subgens).

I'll try to drop a PR tomorrow or during the week-end if I have time.

**Comment by deepu105** [comment]
Hi Mathieu, I would prefer to keep the import JDL tests as it also tests
the subgen, which invokes entity subgen and so on after invoking JHCore

Thanks & Regards,
Deepu
